### PR TITLE
fix: set http2 connection windows size to unlimited

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

The default setting of `connection windows size` is equal to `stream windows size`. Assume a `MergeExecutor` had 3 upstream s1, s2, and s3. When s1 and s2 received barriers, they will stop polling upstream to align barriers. But the upstream will keep sending messages until their stream windows are full. Thus the connection buffer might be overwhelmed by the messages from s1 and s2 before the barrier from s3 is sent to downstream. At this time, `MergeExecutor` will never receive a barrier from s3 and the barrier will not be aligned. This is a deadlock.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#4270

### CODE REVIEW

File: src/storage/src/hummock/mod.rs [hunk 1]
```diff
@@ -208,7 +208,7 @@ impl HummockStorage {
 
     async fn get_compaction_group_id(&self, table_id: TableId) -> HummockResult<CompactionGroupId> {
         match tokio::time::timeout(
-            Duration::from_secs(10),
+            Duration::from_secs(200),
```
Comment (TennyZhuang) [inline src/storage/src/hummock/mod.rs:211]: ?

File: src/common/src/config.rs [hunk 2]
```diff
@@ -20,6 +20,8 @@ use serde::{Deserialize, Serialize};
 use crate::error::ErrorCode::InternalError;
 use crate::error::{Result, RwError};
 
+pub const MAX_CONNECTION_WINDOW_SIZE: u32 = (1 << 31) - 1;
```
Comment (BugenZhao) [inline src/common/src/config.rs:23]:
```suggestion
pub const MAX_CONNECTION_WINDOW_SIZE: u32 = i32::MAX as _;
```

File: src/storage/src/hummock/mod.rs (hunk 1 above)
Comment (yuhao-su) [reply src/storage/src/hummock/mod.rs:211]: It should do retry. 10s is too short and will easily timeout. I'll leave it to 10 and let others fix this...

File: src/rpc_client/src/compute_client.rs [hunk 3]
```diff
@@ -37,15 +39,16 @@ pub struct ComputeClient {
 }
 
 impl ComputeClient {
-    pub async fn new(addr: HostAddr) -> Result<Self> {
+    pub async fn with_address(addr: HostAddr) -> Result<Self> {
         let channel = Endpoint::from_shared(format!("http://{}", &addr))?
+            .initial_connection_window_size(MAX_CONNECTION_WINDOW_SIZE)
```
Comment (BugenZhao) [inline src/rpc_client/src/compute_client.rs:44]: What if we forget this configuration for some other client? 🤣 Should we reuse these in client pool? cc @zwang28

File: src/rpc_client/src/compute_client.rs [hunk 4]
```diff
@@ -37,15 +39,16 @@ pub struct ComputeClient {
 }
 
 impl ComputeClient {
-    pub async fn new(addr: HostAddr) -> Result<Self> {
+    pub async fn with_address(addr: HostAddr) -> Result<Self> {
```
Comment (BugenZhao) [inline src/rpc_client/src/compute_client.rs:42]:
```suggestion
    pub async fn new(addr: HostAddr) -> Result<Self> {
```

**Review by BugenZhao — COMMENTED** [review]
LGTM

File: src/storage/src/hummock/mod.rs (hunk 1 above)
Comment (BugenZhao) [reply src/storage/src/hummock/mod.rs]: I've opened an issue before. #4427

File: src/common/src/config.rs (hunk 2 above)
Comment (yuhao-su) [reply src/common/src/config.rs:23]: it is from https://github.com/hyperium/h2/blob/756384f4cdd62bce3af7aa53a156ba2cc557b5ec/src/proto/mod.rs#L33
But seems I can't get it from tonic. So I just copied that.

File: src/rpc_client/src/compute_client.rs (hunk 3 above)
Comment (yuhao-su) [reply src/rpc_client/src/compute_client.rs:44]: Agree.

**Comment by BugenZhao** [comment]
Good work! I'd like to verify it locally tomorrow. 🥰

**Comment by skyzh** [comment]
Will this cause OOM?

**Comment by BugenZhao** [comment]
> Will this cause OOM?

I don't think so. Seems the window will not be allocated at once, and it's also bounded by `stream count * stream window size`, and we keep the stream window size to default (64kB). Considering that a chunk could be larger than 64kB, these changes can be treated as increasing the channel bound from 16 to 17. I think it's acceptable.

https://github.com/singularity-data/risingwave/blob/626a4c5e3308e546055cce9d7a4b18fa3df0be18/src/stream/src/task/mod.rs#L35-L36

BTW, we add a extra capacity of 1024 for every pair of remote channel before #4495. 😄

**Review by BugenZhao — APPROVED** [review]

**Review by liurenjie1024 — APPROVED** [review]

**Comment by jon-chuang** [comment]
> But the upstream will keep sending messages until their stream windows are full. 

We don't limit this with application buffers? I don't think this phenomenon will happen as that means that we do not configure any back pressure into the merge executor?

**Comment by yuhao-su** [comment]
>we did not previously configure any back pressure into the merge executor?
In remote case, we previously hoped the stream window size would do the back pressure and it worked. But such backpressure caused a deadlock because we set an inappropriate connection window size.

>We don't limit this with application buffers?
If you mean to do something like downstream give credits to upstream, we didn't implement that. I personally think the HTTP2 flow control can do the same job here

**Pull request merged**